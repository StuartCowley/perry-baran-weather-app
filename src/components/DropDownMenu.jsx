import React, { useState } from "react";
import PropTypes from "prop-types";
import "../styles/DropDownMenu.css";
import menu from "../images/menu.png";
import OptionsMenu from "./OptionsMenu";
import useDelayedUnmount from "../hooks/useDelayedUnmount";

function DropDownMenu({ handleSearch }) {
  const [visible, setVisible] = useState(false);
  const mount = useDelayedUnmount(visible, 500);

  const handleHamburger = () => {
    setVisible((prev) => !prev);
  };

  return (
    <>
      <button
        type="button"
        className="drop-down-menu-button"
        onClick={handleHamburger}
      >
        <img src={menu} alt="hamburger menu" />
      </button>
      {mount && (
        <div
          className={`drop-down-menu ${
            visible ? "drop-down-menu__visible" : "drop-down-menu__hidden"
          }`}
          data-testid="drop-down-menu"
        >
          <OptionsMenu handleSearch={handleSearch} setVisible={setVisible} />
        </div>
      )}
    </>
  );
}

DropDownMenu.propTypes = {
  handleSearch: PropTypes.func.isRequired,
};

export default DropDownMenu;
