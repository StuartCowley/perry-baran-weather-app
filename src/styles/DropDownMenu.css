.drop-down-menu-button {
  all: unset;
  cursor: pointer;
}

.drop-down-menu-button > img {
  aspect-ratio: 1;
  width: 40px;
}

.drop-down-menu {
  position: absolute;
  top: var(--top-bar-height);
  right: 0;
  overflow: hidden;
}

.drop-down-menu > * {
  position: relative;
  top: 0;
}

.drop-down-menu__visible > * {
  animation-name: enter;
  animation-duration: 500ms;
}

.drop-down-menu__hidden > * {
  animation-name: exit;
  animation-duration: 500ms;
}

@keyframes enter {
  from {
    top: -200px;
  }
  to {
    top: 0;
  }
}

@keyframes exit {
  from {
    top: 0;
  }
  to {
    top: -200px;
  }
}