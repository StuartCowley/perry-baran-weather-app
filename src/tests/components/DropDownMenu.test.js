import React from "react";
import { render, fireEvent } from "@testing-library/react";
import DropDownMenu from "../../components/DropDownMenu";

describe("DropDownMenu", () => {
  const handleSearch = jest.fn();
  let screen;

  beforeEach(() => {
    screen = render(<DropDownMenu handleSearch={handleSearch} />);
  });

  test("snapshot", () => {
    const { asFragment } = screen;

    expect(asFragment()).toMatchSnapshot();
  });

  test("image", () => {
    const { getByAltText } = screen;
    const hamburgerImage = getByAltText("hamburger menu");

    expect(hamburgerImage).toHaveAttribute("src", "menu.png");
  });

  test("button", () => {
    const { getAllByRole, queryByTestId } = screen;
    const [hamburgerButton] = getAllByRole("button");
    let dropDownMenu = queryByTestId("drop-down-menu");

    expect(hamburgerButton).toHaveAttribute("type", "button");
    expect(dropDownMenu).not.toBeInTheDocument();
    fireEvent.click(hamburgerButton);
    dropDownMenu = queryByTestId("drop-down-menu");
    expect(dropDownMenu).toBeInTheDocument();
  });
});
