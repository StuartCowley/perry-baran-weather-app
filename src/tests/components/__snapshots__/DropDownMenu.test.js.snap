// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropDownMenu snapshot 1`] = `
<DocumentFragment>
  <button
    class="drop-down-menu-button"
    type="button"
  >
    <img
      alt="hamburger menu"
      src="menu.png"
    />
  </button>
</DocumentFragment>
`;
